### PR TITLE
feat: support cyclic foreign keys

### LINKED ISSUE #1717: Cyclic dependency checker should consider different 'as' parameter
I have following example:

```js
var Sequelize = require('sequelize');
var sequelize = new Sequelize('sequelize-test', 'root', '');
var Content = sequelize.define('Content', {});
var Version = sequelize.define('Version', {});

Content.hasMany(Version);
Version.belongsTo(Content);

Content.belongsTo(Version, {
    as: 'ReleaseVersion',
    foreignKey: 'ReleaseVersionId'
});

Version.hasOne(Content, {
    as: 'ReleaseVersion',
    foreignKey: 'ReleaseVersionId'
});

sequelize.sync({ force: true });
```

Which produces following error:

```
Error: Cyclic dependency found. 'Contents' is dependent of itself.
Dependency Chain: Contents -> Versions => Contents
```

I want to associate the active release of a content by the `Content` model, for this I added a new association that links directly to my `Version` model. The key for this id (ReleaseVersionId) should be in the `Content` Table. This key can be null if no version was released. Since Version 2.0.0-dev11 I get an error when constructing something like this. Do I miss something or can I model something like this in a sequelize way?

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Closes #1717

Note that this PR targets v6, I'll bring it into v7 once merged.

This PR fixes the support for circular foreign key references. (e.g. model A that has a foreign key referencing model B, while B also has a foreign key referencing model A).

This was previously causing `Sequelize#sync`, `Sequelize#drop` and `Sequelize#truncate` to throw because:

- `Sequelize#sync` tried to create foreign keys before the target table was created. The updated version still does by default, but splits generation in two steps (table creation, then foreign key constraint creation) if there is a cyclic reference.
- `Sequelize#drop` tried to drop tables that were referenced by other tables. The updated version still does by default but if there is a cyclic reference, it will first remove all foreign keys before dropping all tables.
- `Sequelize#truncate` tried to delete rows that were referenced by other tables. This is left unchanged, as the solution is to use the `cascade` option. (for dialects that support CASCADE)

## TODO

- [x] Add typings for the new methods
- [x] Ensure integration tests are OK

### CODE REVIEW

**Comment by jccpro** [comment]
Nice, we have exactly this problem!

**Review by WikiRik — APPROVED** [review]
Two comments, but looks good!

File: src/dialects/mysql/query-generator.js [hunk 1]
```diff
@@ -169,6 +169,23 @@ class MySQLQueryGenerator extends AbstractQueryGenerator {
     return `${query};`;
   }
 
+  tableExistsQuery(table) {
+    const tableName = table.tableName || table;
+    const schemaName = table.schema;
+
+    let sql = `SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE' AND TABLE_NAME = ${this.escape(tableName)}`;
+
+    // TODO: This replicates showTablesQuery, but we should determine the default schema name instead.
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/mysql/query-generator.js:178]: Yes

File: dev/mssql/2019/start.sh [hunk 2]
```diff
@@ -11,6 +11,6 @@ docker-compose -p sequelize-mssql-2019 up -d
 docker exec sequelize-mssql-2019 \
   /opt/mssql-tools/bin/sqlcmd -S localhost -U SA -P "Password12!" -Q "CREATE DATABASE sequelize_test; ALTER DATABASE sequelize_test SET READ_COMMITTED_SNAPSHOT ON;"
 
-node check.js
+DIALECT=mssql node check.js
```
[the remark this answers is not in the document]
Comment (ephys) [reply dev/mssql/2019/start.sh:14]: It's a change we made in main a while ago, but was missing in v6 (https://github.com/sequelize/sequelize/blob/main/dev/mssql/2019/start.sh)

We can change it but it's how the other dialects are doing it so I opted to replicate it in this PR https://github.com/sequelize/sequelize/blob/main/dev/mariadb/10.3/start.sh#L14

**Comment by WikiRik** [comment]
One test seems to remain broken;
```
  1) [MYSQL] Include
       findAll
         should support including date fields, with the correct timeszone:
     Table 'sequelize_test.users' doesn't exist
  Error
      at Query.run (lib/dialects/mysql/query.js:52:25)
      at /home/runner/work/sequelize/sequelize/lib/sequelize.js:311:28
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async MySQLQueryInterface.showIndex (lib/dialects/abstract/query-interface.js:254:12)
      at async Function.sync (lib/model.js:988:29)
      at async Sequelize.sync (lib/sequelize.js:373:9)
      at async Context.<anonymous> (test/integration/include/findAll.test.js:1563:7)
```
@ephys can you check this out?

**Comment by ephys** [comment]
Yep :)

**Review by WikiRik — APPROVED** [review]

**Comment by WikiRik** [comment]
@ephys I'll leave the merging up to you as it will create a new release

**Comment by ephys** [comment]
Here we go then :)

**Comment by samiurprapon** [comment]
With this support do I have to add `constraints: false`  ? 

Reference : [Documentation](https://sequelize.org/docs/v6/other-topics/constraints-and-circularities/#:~:text=In%20order%20to-,alleviate%20that%2C,-we%20can%20pass)

**Pull request merged**